src/platform/ESP32/OTAImageProcessorImpl.cpp

```diff
@@ -123,6 +134,145 @@ CHIP_ERROR OTAImageProcessorImpl::ProcessBlock(ByteSpan & block)
     return CHIP_NO_ERROR;
 }
 
+#ifdef CONFIG_ENABLE_DELTA_OTA
+static bool verify_chip_id(void * bin_header_data)
+{
+    esp_image_header_t * header = (esp_image_header_t *) bin_header_data;
+    if (header->chip_id != CONFIG_IDF_FIRMWARE_CHIP_ID)
+    {
+        ESP_LOGE(TAG, "Mismatch chip id, expected %d, found %d", CONFIG_IDF_FIRMWARE_CHIP_ID, header->chip_id);
+        return false;
+    }
+    return true;
+}
+
+static bool verify_patch_header(void * img_hdr_data)
+{
+    const uint32_t esp_delta_ota_magic = 0xfccdde10;
+    if (!img_hdr_data)
+    {
+        return false;
+    }
+    uint32_t recv_magic = *(uint32_t *) img_hdr_data;
+    uint8_t * digest    = (uint8_t *) ((uint8_t *) img_hdr_data + 4);
+    if (recv_magic != esp_delta_ota_magic)
+    {
+        ESP_LOGE(TAG, "Invalid magic word in patch");
+        return false;
+    }
+    uint8_t sha_256[DIGEST_SIZE] = { 0 };
+    esp_partition_get_sha256(esp_ota_get_running_partition(), sha_256);
+    if (memcmp(sha_256, digest, DIGEST_SIZE) != 0)
+    {
+        ESP_LOGE(TAG, "SHA256 of current firmware differs from than in patch header. Invalid patch for current firmware");
+        return false;
+    }
+    return true;
+}
+
+esp_err_t verify_header_data(const uint8_t * buf_p, size_t size, int * index)
```

Comment:
I think this method will only verify the first image received. Once patch_header_verified is set to true, any future calls will skip the verification and return ESP_OK. If an image passes this check but isn't applied (e.g. because it fails another check or the OTA provider sends a Discontinue action), then the next image will skip this check because patch_header_verified remains true.

Reply:
Did cleanup if the OTA is aborted or completed